### PR TITLE
feature/train-test-split 

### PR DESCRIPTION
**Pull request by mariru**
**Summary:**

    issues: makes progress for #104 Add Train/Test Separation to convolutional_vae example

I added a test_data attribute to the Data class and a test() method to the VariationalInfernce class. The test method evaluates an evaluation metric on the test data. The evaluation metric should be specified in the model (see examples/bayesian_linear_regression_multidimensional.py where the evaluation metric is MSE). If no evaluation metric is specified, test() evaluates the log probability of the model on the test data.

In addition, I added examples/bayesian_linear_regression_multidimensional.py
This is a modified version of examples/bayesian_linear_regression.py which takes multidimensional input vectors and evaluates MSE on the test data.

**Intended Effect:**

Easy evaluation on test data.

**How to Verify:**

Run examples/bayesian_linear_regression_multidimensional.py
Can, also run all older examples, they are not affected by the changes.

**Documentation:**

N/A

**Reviewer Suggestions:**

@dustinvtran

### CODE REVIEW

**Comment by dustinvtran** [comment]
As mentioned in #104, I think this is a good first approach to the training/test split. In general, this would be subsumed under the criticism interface. So we want to think carefully about how to do this so that it generalizes to things like posterior predictive checks and cross validated assessments of model fit. E.g., should testing be part of inference or its own separate object (possibly taking in the fitted model and data as input)?

**Comment by jramapuram** [comment]
From a purely programatic point of view this still does not address the issues with for example:
  - use of scoping w/ prettytensor [i.e. phase is not passed in]
  - building two objects whereby the test object re-uses the underlaying train params.

**Comment by dustinvtran** [comment]
Okay so I thought about this a little more. 

+ I like that a test function must take `sess` as input. The alternative is to have inference return all learned parameters as NumPy arrays, and then the user would somehow have to wield these NumPy arrays for evaluation at test time. That sounds not very feasible.

Your structure also motivates a lot of things. So here are additional proposals:

+ There will be a `criticism` object. It can do test set and cross validated evaluations, or a PPC.
+ The input is `model`, `variational`, `data` (test or training data), and `sess`. Some of these inputs are optional and more may be necessary. For example, prior predictive checks don't require any inference; therefore it can use its own session and does not have `variational` as input. Predictive checks will also require a discrepancy function `t()` as input.
+ Evaluation metrics return a scalar. PPCs return a reference distribution and realized discrepancy. Graphical checks return a plot.
+ The criticism requires a method from the `model` object. What is the model method's inputs and outputs? 

  Evaluation metrics. The inputs are the usual inputs to `log_prob()`, i.e., data `xs` and latent variables `zs`. The output is the mean of the likelihood for each set of latent variables. It is a TensorFlow vector `[ E_{p(xs | zs[1, :])} [xs], ..., E_{p(xs | zs[S, :])} [xs] ]`.

  PPC's. The inputs are the size of the replicated data set `size`, and a `S x d` matrix of latent variables `zs`. The output is a list of replicated data sets from the likelihood, `[x^{rep, 1}, ..., x^{rep, S}]`, where `x^{rep, s} ~ p(x | zs[s, :])` and `x^{rep, s}` has `size`-many data points. It is a list of objects where the type of the object is the same as the type of the test data.

Let's use your example as reference. We want to calculate mean squared error on held-out data. We will require the model to have a method for evaluation metrics as explained above. The `criticism` object does the following: 1. sample a mini-batch of latent variables from the variational distribution; 2. calculate a vector of likelihood predictions, one element for each sample of latent variables; 3. take the average of this vector, which is our Monte Carlo estimate of the posterior predictive mean; 4. calculate the MSE of this value.

Here are open questions:

+ Should criticism be a class that has a bunch of methods we can call for each criticism we want to do? Or should it be a bunch of standalone functions that we call?
+ What do we call the model method? For evaluation metrics, I think it makes sense to call it `predict()`. For PPC's, I think it makes sense to call it `sample_lik()`.
+ How can do we do criticism during inference? For example, what if want to we print out the test loss as a function of iterations in the algorithm? No clue; may for now users will have to do this in Native TensorFlow similar to the mixture density network example.

Thoughts much appreciated.

**Comment by dustinvtran** [comment]
Following the above proposal, it also extends to @jramapuram's points. Namely, PrettyTensor test scoping will be inside the criticism object and we can add PrettyTensor training scoping inside the inference object. The criticism object can also can re-use the underlying train params because it has access to everything (`model`, `variational`, `data`).
